### DISCUSSION TITLE
xDrip produces incorrect battery data in Nightscout devicestatus uploader JSON object #2501

### DISCUSSION THREAD

**Discussion by dv1** [NightscoutFoundation/xDrip#2503] [General]
The existing code produces a `battery` value like:

    "battery": "50 days (voltage: 285/275)"

This can be found [here](https://github.com/NightscoutFoundation/xDrip/blob/master/app/src/main/java/com/eveningoutpost/dexdrip/UtilityModels/NightscoutBatteryDevice.java#L126) and was introduced by commit 92956f73f49dc199e2539ba055ec0d647876b91e .

However, this is wrong. The Nightscout API docs state that `battery` must be an integer.

Because of this type mismatch, AndroidAPS crashes when its NSClient BG source plugin is read. When it reads the `devicestatus` objects, the parser throws an exception when it tries to parse `battery` as an integer. (AndroidAPS source is [here](https://github.com/nightscout/AndroidAPS/blob/dev/plugins/main/src/main/java/info/nightscout/plugins/sync/nsclient/services/NSClientService.kt#L511), and its schema is [here](https://github.com/nightscout/AndroidAPS/blob/dev/core/ns-sdk/src/main/java/info/nightscout/sdk/remotemodel/RemoteDeviceStatus.kt#L54).)

This needs to be fixed to match the expected type.

Disabling the "Upload OB1G5/G6 transmitter battery" option is a workaround, but _not_ a proper fix.

EDIT: It was also reported [here in the AAPS issues](https://github.com/nightscout/AndroidAPS/issues/2223).

**Comment by jamorham** [comment]
@jwoglom could you take a look at this report, thanks.

**Comment by Navid200** [comment]
@dv1 Would you please have a look at the following?

Settings &#8722;> Cloud Upload &#8722;>Nightscout Sync &#8722;> Extra Options

Do you have Upload OB1G5/G6 transmitter battery enabled on that page?

**Reply by dv1** [reply]
@Navid200 Yeah, I wrote it already above:

> Disabling the "Upload OB1G5/G6 transmitter battery" option is a workaround, but not a proper fix.

So we can use that for now, but that code needs fixing.

**Reply by Navid200** [reply]
I don't know why you say "we can use that for now".
You enabled that feature.  It's not on by default.

**Reply by dv1** [reply]
I mean we can keep it turned off as a workaround to avoid producing more invalid data.

**Reply by Navid200** [reply]
How can Nightscout tell me what is valid and what is invalid in a new data I am uploading to Nightscout?
We can upload this to an alternative to Nightscout also.  Is Nightscout going to say we are in violation then?

**Comment by Navid200** [comment]
Why is that a problem?  
I may want to use it and I don't use AAPS.  What's wrong with that?

**Reply by dv1** [reply]
The data is invalid. The Nightscout API docs clearly state that the battery value must be an integer. See for example here: https://crek9u5.ns.10be.de:9173/api-docs/  - go to "uploader" below.

**Reply by Navid200** [reply]
But, we created a new field for this.  I don't think Nightscout doc has been updated to reflect this new field, has it?

**Reply by dv1** [reply]
Oh, interesting. So a new field was proposed? Which one? Is this documented somewhere, perhaps on Gitter or Discord?

**Reply by Navid200** [reply]
This is the PR:
https://github.com/NightscoutFoundation/xDrip/pull/1709

**Reply by Navid200** [reply]
@jwoglom Is this something we can change without affecting users who need it and asked for it?

**Reply by dv1** [reply]
Ah, I see. I was unaware of that PR.
This brings up a serious problem - the battery used to be an integer value, now it is a string. This works only if your int parser code is lenient and ignores it when values past the initial digits aren't digits. Example: "38 days (voltage: 307/288)" - lenient parsers will just parse "38" and ignore the invalid rest. The AAPS parser isn't lenient, and rejects this because it is not a valid integer.

**Reply by Navid200** [reply]
Thanks for reporting.
Let's wait for @jwoglom to respond.
If this needs to be edited, we can reopen the issue.

Thanks for reporting.

**Reply by dv1** [reply]
No problem. One last note to my explanation above: Even if a parser is lenient, we unfortunately have another problem - the old battery value (the integer one) was a percentage from what I recall. The new value does not store percentages. So, in the example above, a lenient parser parses the integer "38", which is then interpreted as 38%.

**Reply by Navid200** [reply]
If you disable the new feature, you should get the old value.
If it was a percentage, it will be a percentage.

**Reply by Navid200** [reply]
I understand that this breaks AAPS, It  was not our intent to do that.

But, I asked when this PR was being put together exactly what would happen after this was merged for someone who left this at default.  And I requested that in that case, there should be no change.
I was told that was the case.   That nothing will change if the user does not enable the new feature.

**Reply by dv1** [reply]
Yes, and this makes sense. When I introduce new features in code that is already in use, I make these features opt-in to ensure the existing default behavior remains unchanged. I agree 100% with you there.

Separately from that though this changes the *type* of the battery field from an integer to a string, without any indications anywhere that this happens. And this is not how it should have been done. Sadly, I learned too late of this PR, otherwise I would have mentioned that. A better approach would have been a separate field like "extendedBatteryStatus" that exists next to the battery field. This would have been 100% backwards compatible and not have broken anything.

Now though it is too late, since these string values are present in existing Nightscout DBs already. AAPS will need extra code to check if this is a string or an integer.

**Reply by Navid200** [reply]
Let's see what @jwoglom thinks.
He is not junior like me.

**Reply by jwoglom** [reply]
Coming from the person who implemented this in xDrip: the initial motivation behind using a string for the battery field here is that, at least at the time this was added, it was what the Nightscout UI showed next to the device, and the Nightscout UI didn't care if it was an int or a string. There's also some complexity in G5/G6 transmitters as to what single number should be used or is most important to look at when tracking the battery state. There isn't really a way to calculate a percent battery which is left, unless you divide by the number of total expected days which is just an estimation. And, showing the percent of days left isn't super useful to the user. So for all of these reasons a human readable description which showed the number of days and battery voltage was used here, since it seemed less confusing than an arbitrary battery percent number.

Now, if the schema called for this always being an integer, it sounds like that information was unfortunately missed at the time, although I wonder whether assuming this value was always a number in the JSON was enforced or even written down at all in the first place. (If it was defined as an integer somewhere in a canonical schema, I didn't see it; to be honest, I didn't even know if there _was_ any reference as to the types of these fields besides how the UI parses them.)

It's not necessarily essential that the battery field be a string going forward for a few reasons. One, all of the data within the battery string field (battery days, total days, and voltage) is already present as JSON fields, so removing the details within the string doesn't result in losing any data. Two, the value could be set as an integer which is the number of days remaining divided by total days -- since the latter is usually 100 this would work out okay as a battery estimation in most cases, if Nightscout itself or clients which read from NS (like AAPS) do no special handling. And three, there is an uploader.type == DEXCOM_TRANSMITTER in the JSON which could be used to, in the NS/AAPS UI, display the detailed transmitter information in place of that integer field.

To solve this the "right" way, we could:
* In xDrip, stop populating the string and do an integer as other clients may expect.
* In the Nightscout UI, case on a battery status from a dexcom transmitter and specifically show the details which previously were present in that string rather than the raw integer percent. This ensures the same data is displayed regardless of whether the user has this feature enabled on an xDrip version prior to the above fix or after it.
* In AAPS, add a "type" optional string field to the Uploader data class, as well as optional fields for any of the dexcom transmitter-filled fields, e.g. days, daysEstimate, voltagea, voltageb, and status. Then add some kind of UI check for whether the type is dexcom_transmitter and display these fields, if desired.
* In AAPS, consider updating RemoteDeviceStatus to either have [lenient parsing enabled for the kotlin serializer](https://github.com/Kotlin/kotlinx.serialization/blob/master/docs/json.md#lenient-parsing) (which will allow the battery field to be specified in the schema as an int but ingest a string), or set battery to a JsonPrimitive which can either be an int or string, to work around the existing behavior if people don't upgrade xDrip.

Alternatively, AAPS could work around this and we essentially change the Nightscout schema to either be an int -- representing battery percent -- or a string, for devices like transmitters which don't exactly have the concept, where it denotes a description of the battery state just not in numeric form. This would only require work to fix this from AAPS, since other systems don't seem to have the same degree of type enforcement on these Nightscout fields.

**Reply by Navid200** [reply]
>  Two, the value could be set as an integer which is the number of days remaining divided by total days

I don't think that's a good idea. The number of days is a valuable piece of information.  But, there should not be an association between the number of days and the battery voltages.  Those are different things.   In some cases, they may become related.  But, assuming that they always are is not a good idea.

There are different types of transmitters.  Someone could be using a rechargeable battery.  If the battery voltage drops, there may still be plenty of days left.

On the other hand, you could start using a transmitter that has very recently been manufactured.  In that case, when you reach day 110, you will still have full batteries.  

If what you meant was to convert the number of days to an integer only while leaving the voltages, I see no problem and I have just misunderstood your point.

**Comment by dv1** [comment]
Those are good points. The battery schema was probably designed with assumptions in mind that turned out to not work in reality.

I have a simple fix for AAPS that checks if the battery value is a string. If so, it replaces it with a fixed battery value 100. Not clean, but does not require extensive refactoring, unlike other approaches.

I would not just use a lenient parser though (which inteprets "50 days (voltage: 285/275)" as 50 for example), since this introduces the silent bug that a day count is misinterpreted as a percentage.
